mgosimplekv/kv.go

```diff
@@ -0,0 +1,171 @@
+// database. The collection's underlying session must be closed when the
+// query is complete.
+func (s *kvStore) c(ctx context.Context) *mgo.Collection {
+	return s.coll.With(s.session(ctx))
+}
+
+type kvDoc struct {
+	Key    string    `bson:"_id"`
+	Value  []byte    `bson:"value'`
+	Expire time.Time `bson:",omitempty"`
+}
+
+// Get implements simplekv.Store.Get by retrieving the document with
+// the given key from the store's collection.
+func (s *kvStore) Get(ctx context.Context, key string) ([]byte, error) {
+	coll := s.c(ctx)
+	defer coll.Database.Session.Close()
+
+	var doc kvDoc
+	if err := coll.FindId(key).One(&doc); err != nil {
+		if errgo.Cause(err) == mgo.ErrNotFound {
+			return nil, simplekv.KeyNotFoundError(key)
+		}
+		return nil, errgo.Mask(err)
+	}
+	return doc.Value, nil
+}
+
+// Set implements simplekv.Store.Set by upserting the document with
+// the given key, value and expire time into the store's collection.
+func (s *kvStore) Set(ctx context.Context, key string, value []byte, expire time.Time) error {
+	coll := s.c(ctx)
+	defer coll.Database.Session.Close()
+
+	_, err := coll.UpsertId(key, bson.D{{
+		"$set", bson.D{{
+			"value", value,
+		}, {
+			"expire", expire,
+		}},
+	}})
+	return errgo.Mask(err)
+}
+
+var updateStrategy = retry.Exponential{
+	Initial:  time.Microsecond,
+	Factor:   2,
+	MaxDelay: 500 * time.Millisecond,
+	Jitter:   true,
+}
+
+// Update implements simplekv.Store.Update.
+func (s *kvStore) Update(ctx context.Context, key string, expire time.Time, getVal func(old []byte) ([]byte, error)) error {
+	coll := s.c(ctx)
+	defer coll.Database.Session.Close()
+
+	r := retry.StartWithCancel(updateStrategy, nil, ctx.Done())
+	for r.Next() {
+		var doc kvDoc
+		if err := coll.Find(bson.D{{"_id", key}}).One(&doc); err != nil {
+			if errgo.Cause(err) != mgo.ErrNotFound {
+				return errgo.Mask(err)
+			}
+			newVal, err := getVal(nil)
+			if err != nil {
+				return errgo.Mask(err, errgo.Any)
+			}
+			err = coll.Insert(kvDoc{
+				Key:    key,
+				Value:  newVal,
+				Expire: expire,
+			})
+			if err == nil {
+				return nil
+			}
+			if !mgo.IsDup(err) {
+				return errgo.Mask(err)
+			}
+			// A new document has been inserted after we did the FindId and before Insert,
+			// so try again.
+			continue
+		}
+		newVal, err := getVal(doc.Value)
+		if err != nil {
+			return errgo.Mask(err, errgo.Any)
+		}
+		if bytes.Equal(newVal, doc.Value) {
+			return nil
+		}
+		err = coll.Update(bson.D{{
+			"_id", key,
+		}, {
+			"value", doc.Value,
+		}}, bson.D{{
+			"$set", bson.D{{
+				"value", newVal,
+			}, {
+				"expire", expire,
+			}},
+		}})
+		if err == nil {
+			return nil
+		}
+		if err != mgo.ErrNotFound {
+			return errgo.Mask(err)
+		}
+		// The document has been removed or updated since we retrieved it,
+		// so try again.
+	}
+	if r.Stopped() {
+		return errgo.Notef(ctx.Err(), "cannot update key")
+	}
+	return errgo.Newf("too many retry attempts trying to update key")
```

Comment:
I still think this might be worth making an error code for. Leave it for a later PR though.